com.unity.netcode.gameobjects/Runtime/Core/PackageMetadata.cs

```diff
@@ -0,0 +1,24 @@
+using Unity.Collections;
+using UnityEngine;
+
+namespace Unity.Netcode
+{
+    public class PackageMetadata : ScriptableObject
```

Comment:
I have 2 topics in my mind:

#### Topic 1

So, reading the comment inline above (there might not be a NetworkManager serialized to disk/scene/asset all the time), I started to wonder since we're creating this ScriptableObject type, could we bundle an asset inside the package and read from that directly and read from it instead?

There's Resources folder which we could create and store a config ScriptableObject:
- https://docs.unity3d.com/ScriptReference/Resources.html
- https://blog.theknightsofunity.com/unity-resources-folder-how-to-use-it/

To be completely honest, I don't have a specific solution in mind, I'm just trying to brainstorm to not ILPP it if possible, maybe we'll end up in ILPP land eventually.

#### Topic 2

Also with this PR, we're effectively making the package exact matching string version compatible, which is a significant change, not saying we should not do it but I think we might want to think about it again — more specifically, should the rule be major & minor version changes won't be compatible but patch versions would? (like major.minor.patchX would still be compatible with major.minor.patchY)

Reply:
I actually didn't mean to leave this as a ScriptableObject type... using a ScriptableObject with the value stored in it during its OnValidate was one of the things I tried... the problem is getting access to it after it's written. Forcing use of the resources folder doesn't seem like a good solution because it requires deploying with a resources folder just for this feature... ScriptableObjects can also be in AssetBundles or referenced by GameObjects but again, forcing the user to have an asset bundle we can load or to create a GameObject just to store a reference to our package metadata ScriptableObject doesn't seem like a good thing for our users.

On topic 2, I thought about that, too, but we have already introduced many patch versions that had incompatible message formats. We have no support right now for saying "if 1.0.1 sends a message that's received by 1.0.2, adjust the message in this way to allow 1.0.2 to read it" and we also have no process or tools or even rules to prevent making message incompatibility in patch versions. And I would argue we don't want them, because that limitation would prevent us being able to solve whole classes of bugs without putting out minor versions... a lot of bugs would end up basically going to a graveyard to be resurrected when a new minor version is planned, and then only the ones we have man power at that point to address would get addressed.

I thought about hashing the messages we send to determine if the protocol has actually changed... but that would be near impossible to do correctly since we have things like SceneEventData that aren't message types but are included by them, so we'd have to end up using ILPP and not only recursively checking every field (and its fields (and its fields)) for changes, but we'd also have to check the Serialize and Deserialize methods, and any methods they call, and the Handle method, and any methods it calls, all the way down the stack, with any slight change to any of that triggering a change in the hash because there'd be no way to reliably heuristically determine if they change the serialized data format.

Ultimately I do agree that requiring patch versions to match might be overkill in some cases, it's the only solution I could think of that wouldn't have any false negatives with our current approach to updates and bugfixes with no message versioning solutions.

Reply:
> Forcing use of the resources folder doesn't seem like a good solution because it requires deploying with a resources folder just for this feature...

We're not forcing users to do anything, we just add 1 asset to Resources folder inside the package, not even in the project (we can have Resources folder within the package boundaries and that works). Also, why using a well known Unity feature is not a good solution?

> ScriptableObjects can also be in AssetBundles or referenced by GameObjects

and again, we don't need AssetBundles or some GameObjects/Prefabs in scenes etc. — we'd just create an asset inside the package under Resources folder and `Resources.Load` that asset, that's it?

Reply:
On the topic 2, it's unfortunate that in the NGO1 codebase, it is super easy to break the protocol and really hard to track whether it'd be broken or not. So, I'm with you on that, maybe we should just make this PR change a little bit more visible (changelog, an internal slack thread/notification etc) to at least let people know this is coming and here's the rationale behind it.